resolution/design/00_resolution.md

```diff
@@ -0,0 +1,331 @@
+
+By providing the necessary resolution wrapper to user-provided ![S(\mathbb{Q};\mathbf{p})] models
+PACE will enable the routine extraction of physically relevant parameters from measured inelastic neutron scattering data.
+
+## Deliverables
+- modifications to the `OptModel`, `OptFunctions`, and `OptFunction` classes
+- a new `OptFunction` subclass, `TobyfitOptFunction` to perform resolution convolution following the `tobyfit` scheme
+- possibly more `OptFunction` subclasses to perform resolution convolution following other schemes
+
+
+# Existing Solution
+Resolution convolution is accomplished within the existing `multifit` framework through specialisation of the MATLAB class `mfclass`.
+Like [`OptModel`] `mfclass` contains independent-variable array(s), foreground and background function(s) plus their parameters and meta information for fixing and binding-together parameters.
+
+Additionally, `mfclass` supports 'wrapping' its foreground and background functions in arbitrary other functions by including a `mfclass_wrapfun` object.
+When the `mfclass` methods `simulate()` or `fit()` are called the `mfclass_wrapfun` property's `wrap_functions_and_parameters()` method is used to
+
+1. (optionally) initialise any temporarily cached values
+2. construct the 'wrapped' functions which will be called with the independent-variable arrays, the `mfclass` foreground/background function properties, the cached information from the initialisation, and the parameters needed by the function(s)
+
+A diagram showing the relationships between the classes of the `multifit` framework follows
+
+![mfclass]
+
+In the specific case of convoluting the instrumental resolution with one or more user-provided functions, a special class `mfclass_tobyfit` is used which has `mfclass` as its superclass.
+Details required for the resolution estimation and integration are provided as additional properties of each `mfclass_tobyfit` object, and are initialised with sensible default values.
+This enables resolution convoluted fitting to be performed by replacing, e.g., the call to `multfit_sqw()` in the [existing `multifit` solution] with a call to a different gateway function `tobyfit()`:
+
+```matlab
+obj = tobyfit(sqw_obj1 {, sqw_obj2, ...}); % object containing 1+ sqw object(s)
+obj = obj.set_fun(@user_sqw_function, p_i); % define a foreground function
+obj = obj.set_bfun(@user_bkg_function, p_j); % define a background function
+obj = obj.set_free(...); % optionally fix foreground parameters
+obj = obj.set_bfree(...); % optionally fix background parameters
+obj = obj.set_bind(...); % optionally bind foreground parameters
+obj = obj.set_bbind(...); % optionally bind background parameters
+[sim_obj(s), p_fit] = obj.fit();  
+```
+
+The flexibility provided by the interaction of `mfclass`, `mfclass_tobyfit`, `mfclass_wrapfun`, and `mfclass_plist` comes with the significant drawback that the code is difficult to understand and therefore difficult to maintain.
```

Comment:
There is a mistake in the class diagram I produced which has, unfortunately, caused some of your confusion &mdash; `mfclass_plist` does not contain another `mfclass_plist` as its property `plist_`.
Instead `plist_` is the input provided to the class constructor (initially) and can be 'wrapped' to include additional functions and parameters (?). I'm not entirely sure what it is intended for though, perhaps @tgperring can help us both out.

As for your first question:
`mfclass_tobyfit` extends the properties of `mfclass` to include *some* of the information needed to perform the resolution estimate and convolution. I do not believe that any methods of `mfclass` are modified by `mfclass_tobyfit`. Using the `mfclass_tobyfit` constructor directly would not yield a useful convolution object, however, since one must *also* define a special `mfclass_wrapfun` to actually do everything.
The `tobyfit()` gateway function constructs this special `mfclass_wrapfun` object by providing it with a resolution-estimate function and convolution function, both of which are implemented as extension methods for `sqw`. The exact functions provided to the `mfclass_wrapfun` constructor are dependent on the instrument type encoded in the `sqw` objects input to `tobyfit()`.

I can try and expand on this in the design document, but I'm not sure that I can do so in a concise manner.

Reply:
Maybe a short sentence saying that the actual implementation is more complicated and needs overloading both `mfclass` and `mfclass_wrapfun`? (and emphasise that the new proposed design would only involve subclassing a single class?)

Reply:
`mfclass_wrapfun` isn't overloaded, which I think contributes to the difficulty in understanding how `tobyfit()` works.

I hope that by removing `mfclass_wrapfun` and sticking everything in the subclasses of `OptFunction`, e.g., `TobyfitOptFunction`, that following the execution path of the code will be less confusing to new users/developers.
If the user-provided function needs to be 'wrapped', as in `tobyfit`, that should be explicit in the class which implements the `tobyfit` method of resolution convolution.